### ISSUE TITLE
Ability to install multiple instances of a distro with `wsl --install <distro> <instance_name> <location> --version 1/2`

### ISSUE THREAD

**Issue by NotTheDr01ds** [microsoft/WSL#6883]
**Is your feature request related to a problem? Please describe.**

WSL instances are great as "single purpose", sometimes even "throwaway" containers for testing out a piece of functionality.  Today, I do this by keeping "pristine" copies of each distro by `wsl --export`ing them immediately after installation.  I can then spin up new instances by `wsl --import`ing those tarballs.

I realize that it's also possible to get the pristine tarball from the `Program Files\WindowsApps` folder, but this is even more painful, IMHO, since you have to (a) find and get access to the right folder from an elevated PowerShell, and (b) unzip the tarball before `wsl --import` can handle it.

The new `wsl --install <distro>` sounds great (I'm not on Preview), but doesn't sound like it goes far enough towards solving this particular problem.

**Describe the solution you'd like**

Extend `wsl --import <distro>` to accept optional parameters for `<instance_name>`, `<location>`, and `--version 1/2` similar to the `wsl --import` command.

**Describe alternatives you've considered**

Alternatives are listed in the "problem statement" above.

**Additional context**

N/A

**Comment by ad-on-is** [comment]
An alternative would be to `wsl --export <distroname> <tarballlocation>` an installed distro to a specific location, which already saves it as a tarball, and then `wsl --import <distroname_x> <installlocation> <tarballlocation>` it multiple times.

**Comment by Baalkikhaal** [comment]
@NotTheDr01ds is there a difference between multiple instances of a distribution and multiple terminals of a distribution 🤔 ?

**Comment by Biswa96** [comment]
The term "instance" is somewhat wrong in this context. The author means multiple _installation_ of same distribution.  For example, one can install Debian twice with _different_ registered WSL distro names for two different development environment.

**Comment by Baalkikhaal** [comment]
@Biswa96  That is neat 😄 .

# Testing multiple instances/installations

To test it, I did

```bash
C:\Users\fubar>wsl --list --verbose
  NAME            STATE           VERSION
* Ubuntu-18.04    Stopped         2
```

Export the distro to a tar file

```bash
C:\Users\fubar>wsl --export Ubuntu18.04 D:\WSL_Backups\ubuntu_18_lts.tar
```

Import the distro as a new installation/instance

```bash
C:\Users\fubar>wsl --import Ubuntu-copy D:\WSL_Distro_Storage\Ubuntu D:\WSL_Backups\ubuntu_18_lts.tar
```

WSL distro listing

```bash
C:\Users\fubar>wsl --list
Windows Subsystem for Linux Distributions:
Ubuntu-18.04 (Default)
Ubuntu-copy
```

Nautilus is getting listed for both the copies.

![image](https://user-images.githubusercontent.com/8896597/117016342-3c703280-ad10-11eb-926f-773029c0d164.png)

# Concerns

Will the exported backups be successfully imported after there is a `wsl --update` as this might upgrade Linux kernel 🤔 ?

# References:

- `wsl --help`
- [Import Linux distribution as WSL](https://docs.microsoft.com/en-us/windows/wsl/use-custom-distro)

**Comment by Biswa96** [comment]
Updating the Linux kernel does not* affect the installed GNU/Linux distributions in WSL. Linux kernel folder is mounted separately in `/tools` folder by initrd.

<sub><sup>*generally</sup></sub>

**Comment by NotTheDr01ds** [comment]
> An alternative would be to `wsl --export <distroname> <tarballlocation>` an installed distro to a specific location, which already saves it as a tarball, and then `wsl --import <distroname_x> <installlocation> <tarballlocation>` it multiple times.

Yes, that's what I do today as mentioned in the first paragraph in the issue ;-).

**Comment by NotTheDr01ds** [comment]
> The term "instance" is somewhat wrong in this context. The author means multiple _installation_ of same distribution. For example, one can install Debian twice with _different_ registered WSL distro names for two different development environment.

Agreed, "instance" or "installation".  Either is better than the WSL doc/help which calls each installation a "distribution".

**Comment by ohadschn** [comment]
Ideally this would also create Start Menu shortcuts and Windows Terminal integration (specifically conveniently create new terminal windows with specific instances).

**Comment by fischerdouglas** [comment]
I hate to say this... But just give up and use Docker Desktop over WSL2 to create those different instances.

https://github.com/microsoft/WSL/issues/9977#issuecomment-2582515213
